c_binding/lsm_local_disk.c

```diff
@@ -751,3 +804,214 @@ int lsm_local_disk_vpd83_get(const char *disk_path, char **vpd83,
+            rc = LSM_ERR_LIB_BUG;
+            _lsm_err_msg_set(err_msg,
+                             "BUG: VPD page 0x%02x is supported, "
+                             "but failed with error %d(%s)", ioctl_rc,
+                             strerror_r(ioctl_errno, strerr_buff, _LSM_ERR_MSG_LEN));
+            goto out;
+
+        }
+
+        rc = LSM_ERR_NO_SUPPORT;
+
+        _lsm_err_msg_set(err_msg, "SCSI VPD 0x%02x page is not supported",
+                         page_code);
+        goto out;
+    }
+
+ out:
+    if (rc != LSM_ERR_OK)
+        memset(vpd_data, 0, (size_t) vpd_data_len);
+
+    return rc;
+}
+
+/*
+ * Preconditions:
+ *  err_msg != NULL
+ *  sd_path != NULL
+ *  fd != NULL
+ */
+static int _sg_io_ioctl_open(char *err_msg, const char *sd_path, int *fd)
+{
+    int rc = LSM_ERR_OK;
+    char strerr_buff[_LSM_ERR_MSG_LEN];
+
+    assert(err_msg != NULL);
+    assert(sd_path != NULL);
+    assert(fd != NULL);
+
+    *fd = open(sd_path, O_RDONLY|O_NONBLOCK);
+    if (*fd < 0) {
+        switch(errno) {
+        case ENOENT:
+            rc = LSM_ERR_NOT_FOUND_DISK;
+            _lsm_err_msg_set(err_msg, "Disk %s not found", sd_path);
+            goto out;
+        case EACCES:
+            rc = LSM_ERR_PERMISSION_DENIED;
+            _lsm_err_msg_set(err_msg, "Permission denied: Cannot open %s "
+                             "with O_RDONLY and O_NONBLOCK flag", sd_path);
+            goto out;
+        default:
+            rc = LSM_ERR_LIB_BUG;
+            _lsm_err_msg_set(err_msg, "BUG: Failed to open %s, error: %d, %s",
+                             sd_path, errno,
+                             strerror_r(errno, strerr_buff, _LSM_ERR_MSG_LEN));
+        }
+    }
+ out:
+    return rc;
+}
+
+int lsm_local_disk_rpm_get(const char *sd_path, int32_t *rpm,
+                           lsm_error **lsm_err)
+{
+    uint8_t vpd_data[_T10_SBC_VPD_BLK_DEV_CHA_MAX_PAGE_LEN];
+    int fd = -1;
+    char err_msg[_LSM_ERR_MSG_LEN];
+    int rc = LSM_ERR_OK;
+    struct t10_sbc_vpd_bdc *bdc = NULL;
+
+    rc = _check_null_ptr(err_msg, 3 /* arg_count */, sd_path, rpm, lsm_err);
+    if (rc != LSM_ERR_OK) {
+        goto out;
+    }
+
+    _lsm_err_msg_clear(err_msg);
+
+    rc = _sg_io_ioctl_open(err_msg, sd_path, &fd);
+    if (rc != LSM_ERR_OK)
+        goto out;
+
+    rc = _sg_io_ioctl_vpd(err_msg, fd, _T10_SBC_VPD_BLK_DEV_CHA,  vpd_data,
+                          _T10_SBC_VPD_BLK_DEV_CHA_MAX_PAGE_LEN);
+    if (rc != LSM_ERR_OK)
+        goto out;
+    bdc = (struct t10_sbc_vpd_bdc *) vpd_data;
+    if (bdc->pg_code != _T10_SBC_VPD_BLK_DEV_CHA) {
+        rc = LSM_ERR_LIB_BUG;
+        _lsm_err_msg_set(err_msg, "Got corrupted SCSI SBC "
+                         "Device Characteristics VPD page, expected page code "
+                         "is %d but got % " PRIu8 "",
+                         _T10_SBC_VPD_BLK_DEV_CHA, bdc->pg_code);
+        goto out;
+    }
+
+    *rpm = be16toh(bdc->medium_rotation_rate_be);
```

Comment:
I was thinking more like:

`be16toh(*((uint16_t *)&(vpd_data[4])));`

But having the structure works too.